File: src/main/java/net/b0n541/pmcts/game/skat/SkatGameState.java
//package net.b0n541.pmcts.game.skat;
//
//import net.b0n541.pmcts.mcts.GameMove;
//import net.b0n541.pmcts.mcts.GameState;
//
//import java.util.ArrayList;
//import java.util.List;
//import java.util.Map;
//
//public class SkatGameState implements GameState<SkatGameMove> {
//
//    private final SkatGameType gameType;
//    private final PlayerPosition gameForeHand;
//    private final List<SkatGameMove> gameMoves;
//    private final List<Trick> tricks;
//    private final PlayerPosition nextPlayer;
//
//    public SkatGameState(final SkatGameType gameType, final PlayerPosition gameForeHand) {
//        this.gameType = gameType;
//        this.gameForeHand = gameForeHand;
//        nextPlayer = gameForeHand;
//        gameMoves = new ArrayList<>();
//        tricks = new ArrayList<>();
//        //tricks.add(new Trick(gameForeHand));
//    }
//
//    @Override
//    public String getNextPlayer() {
//        return nextPlayer.toString();
//    }
//
//    @Override
//    public List<String> getPlayers() {
//        return List.of(PlayerPosition.FOREHAND.toString(), PlayerPosition.MIDDLEHAND.toString(), PlayerPosition.REARHAND.toString());
//    }
//
//    @Override
//    public List<SkatGameMove> getPossibleMoves() {
//        return null;
//    }
//
//    @Override
//    public SkatGameState addMove(final SkatGameMove move) {
////        SkatGameState newState = withGameMoves(new ArrayList<>(gameMoves));
////        newState.gameMoves.add(move);
////        final Trick lastTrick = newState.getLastTrick();
////        lastTrick.addMove(move);
////        if (lastTrick.isFinished()) {
////            newState = newState.withNextPlayer(lastTrick.getTrickWinner());
////            newState = newState.withTricks(new ArrayList<>(tricks));
////            newState.tricks.add(new Trick(lastTrick.getTrickWinner()));
////        } else {
////            newState = newState.withNextPlayer(move.player.getNextPlayer());
////        }
////        return newState;
//
//
////        final SkatGameStateBuilder newState = SkatGameState.builder()
////                .gameType(gameType)
////                .gameForeHand(gameForeHand)
////                .gameMoves(new ArrayList<>(gameMoves));
//////        getLastTrick().copy()
////        final List<Trick> newTricks = new ArrayList<>(tricks);
////        final Trick newLastTrick = newTricks.get(newTricks.size() - 1);
////        newLastTrick.addMove(move);
////
////        if (newLastTrick.isFinished()) {
////            //newTricks.add(new Trick(newLastTrick.getTrickWinner()));
////            newState.nextPlayer(newLastTrick.getTrickWinner());
////        } else {
////            newState.nextPlayer(move.player.getNextPlayer());
////        }
////
////        return newState.build();
//        return null;
//    }
//
//    private Trick getLastTrick() {
//        return tricks.get(tricks.size() - 1);
//    }
//
//    @Override
//    public boolean isGameFinished() {
//        return tricks.size() == 10 && tricks.get(9).isFinished();
//    }
//
//    @Override
//    public Map<String, Double> getGameValues() {
//        return null;
//    }
//
//    @Override
//    public GameMove getLastMove() {
//        return null;
//    }
//}

File: src/main/kotlin/net/b0n541/pmcts/game/skat/SkatGameState.kt
package net.b0n541.pmcts.game.skat

import net.b0n541.pmcts.mcts.GameMove
import net.b0n541.pmcts.mcts.GameState

class SkatGameState(
    val gameType: SkatGameType,
    val declarer: PlayerPosition,
    var nextPlayerPosition: PlayerPosition
) :
    GameState<SkatGameMove> {

    private val gameMoves: MutableList<GameMove> = mutableListOf()

    private val skatCards = Hand()
    private val playerCards = mapOf(
        PlayerPosition.FOREHAND to Hand(),
        PlayerPosition.MIDDLEHAND to Hand(),
        PlayerPosition.REARHAND to Hand()
    )

    private val tricks: MutableList<Trick> = mutableListOf()

    private var declarerPoints: Int = 0
    private var opponentPoints: Int = 0

    private val lastTrick
        get() = tricks.last()

    init {
        tricks.add(Trick(nextPlayerPosition))
    }

    override fun getPlayers(): List<String> =
        listOf(
            PlayerPosition.FOREHAND.toString(),
            PlayerPosition.MIDDLEHAND.toString(),
            PlayerPosition.REARHAND.toString()
        )

    override fun getNextPlayer(): String {
        return nextPlayerPosition.toString()
    }

    override fun getPossibleMoves(): List<SkatGameMove> {
        val allCards = playerCards[nextPlayerPosition]!!

        return if (lastTrick.leadingCard == null) {
            allCards.openCards.map { SkatGameMove(nextPlayerPosition, it) }
        } else {
            val leadingCard = lastTrick.leadingCard!!
            val trumpCards = playerCards[nextPlayerPosition]!!.getTrumpCards(gameType)

            allCards.openCards
                .filter { it.isAllowedOn(leadingCard, gameType, allCards) }
                .map { SkatGameMove(nextPlayerPosition, it) }
        }
    }

    override fun addMove(move: SkatGameMove): SkatGameState {

        require(nextPlayerPosition == move.player)

        val newState = copy()

        // TODO use return newState.apply { ... }
        newState.gameMoves.add(move)

        var lastTrick = newState.lastTrick
        lastTrick.addMove(move)

        if (lastTrick.isFinished) {
            val trickWinner = lastTrick.trickWinner!!
            if (trickWinner == declarer) {
                newState.declarerPoints += lastTrick.cardValues
            } else {
                newState.opponentPoints += lastTrick.cardValues
            }
            if (newState.tricks.size < 10) {
                newState.tricks.add(Trick(trickWinner))
            }
            newState.nextPlayerPosition = trickWinner
        } else {
            newState.nextPlayerPosition = move.player.nextPlayer
        }

        if (newState.isGameFinished) {
            newState.declarerPoints += skatCards.cardValues
        }

        return newState
    }

    private fun copy(): SkatGameState {
        val newState = SkatGameState(gameType, declarer, nextPlayerPosition)
        newState.gameMoves.clear()
        newState.gameMoves.addAll(gameMoves)
        newState.tricks.clear()
        for (trick: Trick in tricks) {
            newState.tricks.add(trick.copy())
        }
        newState.declarerPoints = declarerPoints
        newState.opponentPoints = opponentPoints
        newState.skatCards.takeCards(skatCards.cards)
        newState.playerCards[PlayerPosition.FOREHAND]!!.takeCards(playerCards[PlayerPosition.FOREHAND]!!.cards)
        newState.playerCards[PlayerPosition.MIDDLEHAND]!!.takeCards(playerCards[PlayerPosition.MIDDLEHAND]!!.cards)
        newState.playerCards[PlayerPosition.REARHAND]!!.takeCards(playerCards[PlayerPosition.REARHAND]!!.cards)
        return newState
    }

    override fun isGameFinished(): Boolean {
        return tricks.size == 10 && lastTrick.isFinished
    }

    override fun getGameValues(): Map<String, Double> {
        return mapOf(
            declarer.toString() to declarerPoints.toDouble(),
            declarer.nextPlayer.toString() to opponentPoints.toDouble(),
            declarer.nextPlayer.nextPlayer.toString() to opponentPoints.toDouble()
        )
    }

    override fun getLastMove(): GameMove = gameMoves.last()

    fun dealPlayerCards(player: PlayerPosition, cards: List<Card>) {
        playerCards[player]!!.takeCards(cards)
    }

    fun dealSkat(cards: List<Card>) {
        skatCards.takeCards(cards)
    }
}